File: TelerikUIExamples/TelerikUIExamples/DataFormLabelAlignment.m
//
//  DataFormLabelAlignment.m
//  TelerikUIExamples
//
//  Copyright © 2015 Telerik. All rights reserved.
//

#import "DataFormLabelAlignment.h"
#import <TelerikUI/TelerikUI.h>
#import "EmployeeInfo.h"

@interface DataFormLabelAlignment () <TKDataFormDelegate>

@end

@implementation DataFormLabelAlignment
{
    EmployeeInfo *_info;
    TKDataForm *_dataForm;
    TKDataFormEntityDataSource *_dataSource;
    NSString *_alignmentMode;
}

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        [self addOption:@"Top Alignment" selector:@selector(prepareTopAlignment)];
        [self addOption:@"Left Alignment" selector:@selector(prepareLeftAlignment)];
        [self addOption:@"Top Inline Alignment" selector:@selector(prepareTopInlineAlignment)];
        [self addOption:@"Table View Layout" selector:@selector(prepareTableLayout)];
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];

    _dataForm = [[TKDataForm alloc] initWithFrame:self.view.bounds];
    _dataForm.autoresizingMask = UIViewAutoresizingFlexibleHeight | UIViewAutoresizingFlexibleWidth;
    _dataForm.backgroundColor = [UIColor colorWithRed:0.937 green:0.937 blue:0.960 alpha:1.0];
    _dataForm.delegate = self;
    [self.view addSubview:_dataForm];
    
    _info = [[EmployeeInfo alloc] init];
    _dataSource = [[TKDataFormEntityDataSource alloc] initWithObject:_info];
    
    [_dataSource addGroupWithName:@"Personal Info" propertyNames:@[@"givenNames", @"surname", @"gender", @"idNumber", @"dateOfBirth"]];
    [_dataSource addGroupWithName:@"Contact Info" propertyNames:@[@"employeeId", @"phoneNumber"]];
    
    _dataSource[@"gender"].editorClass = [TKDataFormSegmentedEditor class];
    _dataSource[@"gender"].valuesProvider = @[@"Male", @"Female"];
    
    _dataSource[@"idNumber"].editorClass = [TKDataFormNumberEditor class];
    _dataSource[@"employeeId"].editorClass = [TKDataFormNumberEditor class];
    _dataSource[@"phoneNumber"].editorClass = [TKDataFormPhoneEditor class];
    
    _alignmentMode = @"Top";
    _dataForm.dataSource = _dataSource;
}

#pragma mark TKDataFormDelegate

- (void)dataForm:(TKDataForm *)dataForm updateEditor:(TKDataFormEditor *)editor forProperty:(TKEntityProperty *)property
{
    property.hintText = property.displayName;
    if ([_alignmentMode isEqualToString:@"Top"]) {
        [self performTopAlignmentSettingsForEditor:editor property:property];
    }
    else if ([_alignmentMode isEqualToString:@"TopInline"]) {
        [self performTopInlineAlignmentSettingsForEditor:editor property:property];
    }
    else if ([_alignmentMode isEqualToString:@"Left"]) {
        [self performLeftAlignmentSettingsForEditor:editor property:property];
    }
}

- (CGFloat)dataForm:(TKDataForm *)dataForm heightForHeaderInGroup:(NSUInteger)groupIndex
{
    return 40;
}

- (CGFloat)dataForm:(TKDataForm *)dataForm heightForEditorInGroup:(NSUInteger)gorupIndex atIndex:(NSUInteger)editorIndex
{
    if ([_alignmentMode isEqualToString:@"Top"]) {
        if (gorupIndex == 0 && editorIndex == 2) {
            return 20;
        }
        
        return 65;
    }
    
    if ([_alignmentMode isEqualToString:@"TopInline"] && gorupIndex == 0 && editorIndex == 4) {
        return 75;
    }
    
    return 44;
}

- (void)dataForm:(TKDataForm *)dataForm updateGroupView:(TKEntityPropertyGroupView *)groupView forGroupAtIndex:(NSUInteger)groupIndex
{
    groupView.editorsContainer.backgroundColor = [UIColor whiteColor];
    if (![_alignmentMode isEqualToString:@"Table"]) {
        ((TKStackLayout *)groupView.editorsContainer.layout).spacing = 7;
    }
}


- (void)dataForm:(TKDataForm *)dataForm didSelectEditor:(TKDataFormEditor *)editor forProperty:(TKEntityProperty *)property
{
    UIColor *borderColor = [UIColor colorWithRed:0.000f green:0.478f blue:1.000f alpha:1.00f];
    CALayer *layer = editor.editor.layer;

    if ([editor isKindOfClass:[TKDataFormDatePickerEditor class]]) {
        TKDataFormDatePickerEditor *dateEditor = (TKDataFormDatePickerEditor *)editor;
        layer = dateEditor.editorValueLabel.layer;
    }

    UIColor *currentBorderColor = [UIColor colorWithCGColor:layer.borderColor];
    layer.borderColor = borderColor.CGColor;
    CABasicAnimation *animate = [CABasicAnimation animationWithKeyPath:@"borderColor"];
    animate.fromValue = currentBorderColor;
    animate.toValue = borderColor;
    animate.duration = 0.4;
    [layer addAnimation:animate forKey:@"borderColor"];
}

- (void)dataForm:(TKDataForm *)dataForm didDeselectEditor:(TKDataFormEditor *)editor forProperty:(TKEntityProperty *)property
{
    if ([editor isKindOfClass:[TKDataFormDatePickerEditor class]]) {
        TKDataFormDatePickerEditor *dateEditor = (TKDataFormDatePickerEditor *)editor;
        dateEditor.editorValueLabel.layer.borderColor = [UIColor colorWithRed:0.784f green:0.780f blue:0.800f alpha:1.00f].CGColor;
    }
    editor.editor.layer.borderColor = [UIColor colorWithRed:0.880f green:0.880f blue:0.880f alpha:1.00f].CGColor;
}

#pragma mark - Events

- (void)prepareTopAlignment
{
    _alignmentMode = @"Top";
    [_dataForm reloadData];
}

- (void)prepareLeftAlignment
{
    _alignmentMode = @"Left";
    [_dataForm reloadData];
}

- (void)prepareTopInlineAlignment
{
    _alignmentMode = @"TopInline";
    [_dataForm reloadData];
}

- (void)prepareTableLayout
{
    _alignmentMode = @"Table";
    [_dataForm reloadData];
}

#pragma mark - Style methods

- (void)performTopAlignmentSettingsForEditor:(TKDataFormEditor *)editor property:(TKEntityProperty *)property
{
    editor.style.separatorColor = nil;
    editor.textLabel.font = [UIFont systemFontOfSize:15];
    editor.style.insets = UIEdgeInsetsMake(1, editor.style.insets.left, 5, editor.style.insets.right);

    if (![property.name isEqualToString:@"gender"]) {
        TKGridLayout *gridLayout = editor.gridLayout;
        TKGridLayoutCellDefinition *editorDef = [gridLayout definitionForView:editor.editor];
        editorDef.row = @1;
        editorDef.column = @1;
        
        if ([property.name isEqualToString:@"dateOfBirth"]) {
            TKDataFormDatePickerEditor *dateEditor = ((TKDataFormDatePickerEditor *)editor);
            TKGridLayoutCellDefinition *labelDef = [gridLayout definitionForView:dateEditor.editorValueLabel];
            labelDef.row = @1;
            labelDef.column = @1;
        }

        TKGridLayoutCellDefinition *feedbackLabelDef = [gridLayout definitionForView:editor.feedbackLabel];
        feedbackLabelDef.row = @2;
        feedbackLabelDef.column = @1;
        feedbackLabelDef.columnSpan = 1;

        [self setEditorStyle:editor];
    }
}

- (void)performLeftAlignmentSettingsForEditor:(TKDataFormEditor *)editor property:(TKEntityProperty *)property
{
    editor.style.separatorColor = nil;
    editor.style.insets = UIEdgeInsetsMake(6, editor.style.insets.left, 6, editor.style.insets.right);
    if (![property.name isEqualToString:@"gender"]) {
        [self setEditorStyle:editor];
    }
}

- (void)performTopInlineAlignmentSettingsForEditor:(TKDataFormEditor *)editor property:(TKEntityProperty *)property
{
    editor.style.separatorColor = nil;
    editor.style.insets = UIEdgeInsetsMake(6, editor.style.insets.left, 6, editor.style.insets.right);
    TKGridLayout *gridLayout = editor.gridLayout;

    [self setEditorStyle:editor];

    if ([property.name isEqualToString:@"dateOfBirth"]) {
        TKDataFormDatePickerEditor *dateEditor = ((TKDataFormDatePickerEditor *)editor);
        TKGridLayoutCellDefinition *labelDef = [gridLayout definitionForView:dateEditor.editorValueLabel];
        labelDef.row = @1;
        labelDef.column = @1;
        [gridLayout setHeight:30 forRow:[labelDef.row integerValue]];
        
        TKGridLayoutCellDefinition *feedbackLabelDef = [gridLayout definitionForView:dateEditor.feedbackLabel];
        feedbackLabelDef.row = @2;
        feedbackLabelDef.column = @1;
        feedbackLabelDef.columnSpan = 1;
    }
    
    if ([editor isKindOfClass:TKDataFormTextFieldEditor.class]) {
        editor.style.textLabelDisplayMode = TKDataFormEditorTextLabelDisplayModeHidden;
        TKGridLayoutCellDefinition *titleDef = [gridLayout definitionForView:editor.textLabel];
        [gridLayout setWidth:0 forColumn:[titleDef.column integerValue]];
    }
}

- (void)setEditorStyle:(TKDataFormEditor*)editor
{
    if (editor.selected) {
        return;
    }
    
    CALayer *layer = nil;
    
    if ([editor isKindOfClass:[TKDataFormDatePickerEditor class]]) {
        TKDataFormDatePickerEditor *dateEditor = (TKDataFormDatePickerEditor *)editor;
        layer = dateEditor.editorValueLabel.layer;
        dateEditor.editorValueLabel.textInsets = UIEdgeInsetsMake(0, 7, 0, 0);
        dateEditor.showAccessoryImage = NO;
        dateEditor.accessoryImageView.image = nil;
    }
    else if ([editor isKindOfClass:[TKDataFormTextFieldEditor class]]) {
        layer = editor.editor.layer;
        ((TKTextField *)editor.editor).textInsets = UIEdgeInsetsMake(0, 7, 0, 0);
    }
    
    layer.borderColor = [UIColor colorWithRed:0.880f green:0.880f blue:0.880f alpha:1.00f].CGColor;
    layer.borderWidth = 1.0;
    layer.cornerRadius = 4;
}

@end

File: Examples/Examples/DataForm/DataFormLabelAlignment.cs
using System;
using TelerikUI;
using Foundation;
using ObjCRuntime;
using UIKit;
using CoreAnimation;

namespace Examples
{
	[Register("DataFormLabelAlignment")]
	public class DataFormLabelAlignment : XamarinExampleViewController
	{
		TKDataForm dataForm;
		TKDataFormEntityDataSourceHelper dataSource;
		DataFormAlignmentDelegate dataFormDelegate;
		EmployeeInfo info;
		public string alignment = "Top";

		public override void ViewDidLoad ()
		{
			AddOption ("Top Alignment", PrepareTopAlignment);
			AddOption ("Left Alignment", PrepareLeftAlignment);
			AddOption ("Top Inline Alignment", PrepareTopInlineAlignment);
			AddOption ("Table View Layout", PrepareTableLayout);

			base.ViewDidLoad ();

			this.info = new EmployeeInfo ();
			this.dataSource = new TKDataFormEntityDataSourceHelper (this.info);
			this.dataFormDelegate = new DataFormAlignmentDelegate (this);

			this.dataForm = new TKDataForm (this.View.Bounds);
			this.dataForm.AutoresizingMask = UIViewAutoresizing.FlexibleWidth | UIViewAutoresizing.FlexibleHeight;
			this.dataForm.BackgroundColor = new UIColor (0.937f, 0.937f, 0.960f, 1.0f);
			this.dataForm.Delegate = this.dataFormDelegate;
			this.View.AddSubview (this.dataForm);

			this.dataSource.AddGroup ("Personal Info", new string[] {
				"GivenNames",
				"Surname",
				"Gender",
				"IdNumber",
				"DateOfBirth"
			});

			this.dataSource.AddGroup ("Contact Info", new string[] { "EmployeeId", "PhoneNumber" });

			this.dataSource ["Gender"].EditorClass = new Class ("TKDataFormSegmentedEditor");
			this.dataSource ["Gender"].ValuesProvider = NSArray.FromStrings (new string[] { "Male", "Female" });

			this.dataSource ["IdNumber"].EditorClass = new Class ("TKDataFormNumberEditor");
			this.dataSource ["EmployeeId"].EditorClass = new Class ("TKDataFormNumberEditor");
			this.dataForm.WeakDataSource = this.dataSource.NativeObject;
		}
			
		private void PrepareTopAlignment ()
		{
			this.alignment = "Top";
			this.dataForm.ReloadData ();
		}
			
		private void PrepareLeftAlignment () 
		{
			this.alignment = "Left";
			this.dataForm.ReloadData ();
		}
			
		private void PrepareTopInlineAlignment ()
		{
			this.alignment = "TopInline";
			this.dataForm.ReloadData ();
		}
			
		private void PrepareTableLayout () 
		{
			this.alignment = "Table";
			this.dataForm.ReloadData ();
		}
	}

	class DataFormAlignmentDelegate: TKDataFormDelegate
	{
		DataFormLabelAlignment owner;

		public DataFormAlignmentDelegate(DataFormLabelAlignment owner) {
			this.owner = owner;
		}

		public override void UpdateEditor (TKDataForm dataForm, TKDataFormEditor editor, TKEntityProperty property)
		{
			property.HintText = property.DisplayName;
			if (owner.alignment == "Top") {
				this.PerformTopAlignmentSettingsForEditor (editor, property);
			} else if (owner.alignment == "TopInline") {
				this.PerformTopInlineAlignmentSettingsForEditor (editor, property);
			} else if (owner.alignment == "Left") {
				this.PerformLeftAlignmentSettingsForEditor (editor, property);
			}
		}

		public override nfloat HeightForHeader (TKDataForm dataForm, uint groupIndex)
		{
			return 40;
		}

		public override nfloat HeightForEditor (TKDataForm dataForm, uint groupIndex, uint editorIndex)
		{
			if (owner.alignment == "Top") {
				if (groupIndex == 0 && editorIndex == 2) {
					return 20;
				}

				return 65;
			}

			if (owner.alignment == "TopInline" && groupIndex == 0 && editorIndex == 4) {
				return 75;
			}

			return 44;
		}

		public override void UpdateGroupView (TKDataForm dataForm, TKEntityPropertyGroupView groupView, uint groupIndex)
		{
			groupView.EditorsContainer.BackgroundColor = UIColor.White;
			if (owner.alignment != "Table") {
				((TKStackLayout)groupView.EditorsContainer.Layout).Spacing = 7;
			}
		}

		public override void DidSelectEditor (TKDataForm dataForm, TKDataFormEditor editor, TKEntityProperty property)
		{
			var borderColor = new UIColor (0.000f, 0.478f, 1.000f, 1.00f);
			var layer = editor.Editor.Layer;

			if (editor.IsKindOfClass(new ObjCRuntime.Class(typeof(TKDataFormDatePickerEditor)))) {
				var dateEditor = editor as TKDataFormDatePickerEditor;
				layer = dateEditor.EditorValueLabel.Layer;
			}

			var currentBorderColor = new UIColor(layer.BorderColor);
			layer.BorderColor = borderColor.CGColor;
			var animate = CABasicAnimation.FromKeyPath ("borderColor");
			animate.From = currentBorderColor;
			animate.To = borderColor;
			animate.Duration = 0.4;
			layer.AddAnimation(animate, "borderColor");
		}

		public override void DidDeselectEditor (TKDataForm dataForm, TKDataFormEditor editor, TKEntityProperty property)
		{
			if (editor.IsKindOfClass(new ObjCRuntime.Class(typeof(TKDataFormDatePickerEditor)))) {
				var dateEditor = editor as TKDataFormDatePickerEditor;
				dateEditor.EditorValueLabel.Layer.BorderColor = new UIColor(0.784f, 0.780f, 0.800f, 1.00f).CGColor;
			}
			editor.Editor.Layer.BorderColor = new UIColor(0.880f, 0.880f, 0.880f, 1.00f).CGColor;
		}

		private void PerformTopAlignmentSettingsForEditor(TKDataFormEditor editor, TKEntityProperty property)
		{
			editor.Style.SeparatorColor = null;
			editor.TextLabel.Font = UIFont.SystemFontOfSize (15);
			editor.Style.Insets = new UIEdgeInsets (1, editor.Style.Insets.Left, 5, editor.Style.Insets.Right);
			
			if (property.Name != "Gender") {
				TKGridLayout gridLayout = editor.GridLayout;
				TKGridLayoutCellDefinition editorDef = gridLayout.DefinitionForView (editor.Editor);
				if (editorDef != null) {
					editorDef.Row = new NSNumber (1);
					editorDef.Column = new NSNumber (1);
				}

				if (property.Name == "DateOfBirth") {
					TKDataFormDatePickerEditor dateEditor = (TKDataFormDatePickerEditor)editor;
					TKGridLayoutCellDefinition labelDef = gridLayout.DefinitionForView (dateEditor.EditorValueLabel);
					labelDef.Row = new NSNumber (1);
					labelDef.Column = new NSNumber (1);
				}

				TKGridLayoutCellDefinition feedbackDef = editor.GridLayout.DefinitionForView (editor.FeedbackLabel);
				feedbackDef.Row = new NSNumber (2);
				feedbackDef.Column = new NSNumber (1);
				feedbackDef.ColumnSpan = 1;

				this.SetEditorStyle (editor);
			}
		}

		private void PerformLeftAlignmentSettingsForEditor(TKDataFormEditor editor, TKEntityProperty property)
		{
			editor.Style.SeparatorColor = null;
			editor.Style.Insets = new UIEdgeInsets (6, editor.Style.Insets.Left, 6, editor.Style.Insets.Right);

			if (property.Name != "Gender") {
				this.SetEditorStyle (editor);
			}
		}

		private void PerformTopInlineAlignmentSettingsForEditor(TKDataFormEditor editor, TKEntityProperty property)
		{
			editor.Style.SeparatorColor = null;
			editor.Style.Insets = new UIEdgeInsets (6, editor.Style.Insets.Left, 6, editor.Style.Insets.Right);

			TKGridLayout gridLayout = editor.GridLayout;

			this.SetEditorStyle (editor);

			if (property.Name == "DateOfBirth") {
				TKDataFormDatePickerEditor dateEditor = (TKDataFormDatePickerEditor)editor;
				TKGridLayoutCellDefinition labelDef = gridLayout.DefinitionForView (dateEditor.EditorValueLabel);
				labelDef.Row = new NSNumber (1);
				labelDef.Column = new NSNumber (1);
				gridLayout.SetHeight (44, labelDef.Row.Int32Value);

				TKGridLayoutCellDefinition feedbackDef = editor.GridLayout.DefinitionForView (editor.FeedbackLabel);
				feedbackDef.Row = new NSNumber (2);
				feedbackDef.Column = new NSNumber (1);
				feedbackDef.ColumnSpan = 1;
			}

			if (editor.IsKindOfClass (new Class(typeof(TKDataFormTextFieldEditor)))) {
				editor.Style.TextLabelDisplayMode = TKDataFormEditorTextLabelDisplayMode.Hidden;
				TKGridLayoutCellDefinition titleDef = editor.GridLayout.DefinitionForView (editor.TextLabel);
				gridLayout.SetWidth (0, titleDef.Column.Int32Value);
			}
		}

		void SetEditorStyle(TKDataFormEditor editor) 
		{
			var layer = editor.Editor.Layer;

			if (editor.IsKindOfClass(new ObjCRuntime.Class(typeof(TKDataFormDatePickerEditor)))) {
				var dateEditor = editor as TKDataFormDatePickerEditor;
				layer = dateEditor.EditorValueLabel.Layer;
				dateEditor.EditorValueLabel.Layer.BorderWidth = 1.0f;
				dateEditor.EditorValueLabel.Layer.BorderColor = UIColor.Gray.CGColor;
				dateEditor.ShowAccessoryImage = false;
				dateEditor.EditorValueLabel.TextInsets = new UIEdgeInsets(0, 7, 0, 0);
			}
			else if (editor.IsKindOfClass(new ObjCRuntime.Class(typeof(TKDataFormTextFieldEditor)))) {
				layer = editor.Editor.Layer;
				(editor.Editor as TKTextField).TextInsets = new UIEdgeInsets (0, 7, 0, 0);
			}

			layer.BorderColor = new UIColor (0.880f, 0.880f, 0.880f, 1.00f).CGColor;
			layer.BorderWidth = 1.0f;
			layer.CornerRadius = 4;
		}
	}
}

